packages/api-server/src/price-oracle.ts

```diff
@@ -0,0 +1,281 @@
+    }
+    return Price.from(ckbPrice).toMinFeeRate();
+  }
+
+  private async sendRequest(url: string) {
+    const validateStatus = function (status: number) {
+      return status === 200;
+    };
+    const options = {
+      validateStatus,
+      timeout: POLL_REQUEST_TIME_OUT,
+    };
+
+    let response = await axios.get(url, options).catch((error) => {
+      if (error.response) {
+        // The request was made and the server responded with wrong status code
+        throw new Error(
+          `[${CKBPriceOracle.name}] sendRequest: response failed, statusCode: ${error.response.status}, data: ${error.response.data}, header: ${error.response.headers}`
+        );
+      } else if (error.request) {
+        // The request was made but no response was received
+        // `error.request` is an instance of XMLHttpRequest in the browser and an instance of
+        // http.ClientRequest in node.js
+        throw new Error(
+          `[${
+            CKBPriceOracle.name
+          }] sendRequest: request failed, ${JSON.stringify(error.toJSON())}`
+        );
+      } else {
+        // Something happened in setting up the request that triggered an Error
+        throw new Error(
+          `[${CKBPriceOracle.name}] sendRequest: setup request failed, ${error.message}`
+        );
+      }
+    });
+    return response.data;
+  }
+
+  private async pollPrice(): Promise<string | null> {
+    // rate-limit: 50 requests 1 minute
+    const coingecko = async () => {
+      const tokenId = "nervos-network";
+      const url = `https://api.coingecko.com/api/v3/simple/price?ids=${tokenId}&vs_currencies=usd`;
+      const resObj = await this.sendRequest(url);
+      if (!("usd" in resObj[tokenId])) {
+        throw new Error(
+          `[${CKBPriceOracle.name}] pollPrice: response from ${url} error, result: ${resObj}`
+        );
+      }
+      return new Decimal(resObj[tokenId].usd).toString();
+    };
+
+    // rate-limit: 160000 requests 1 day, 50 requests 10 seconds
+    const binance = async () => {
+      const symbol = "CKBUSDT";
+      const url = `https://api.binance.com/api/v3/trades?symbol=${symbol}&limit=1`;
+      const resObj = await this.sendRequest(url);
+      if (
+        !Array.isArray(resObj) ||
+        resObj.length != 1 ||
+        !("price" in resObj[0])
+      ) {
+        throw new Error(
+          `[${CKBPriceOracle.name}] pollPrice: response from ${url} error, result: ${resObj}`
+        );
+      }
+      return new Decimal(resObj[0]["price"]).toString();
+    };
+
+    // rate-limit: 100 requests per second each
+    const cryptocom = async () => {
+      const symbol = "CKB_USDT";
+      const url = `https://api.crypto.com/v2/public/get-trades?instrument_name=${symbol}`;
+      const resObj = await this.sendRequest(url);
+      if (resObj.code != 0 || !("result" in resObj)) {
+        throw new Error(
+          `[${CKBPriceOracle.name}] pollPrice: response from ${url} error, result: ${resObj}`
+        );
```

Comment:
What about collecting the errors into NewRelic?

Reply:
good idea. how do I call NewRelic?

Reply:
```diff
- .map((p) => (p as PromiseRejectedResult).reason);
+ .map((p) => {
+   const reason = (p as PromiseRejectedResult).reason;
+   newrelic.noticeError(reason);
+   return reason;
+ });
```

After applying the above changes to Line231, NewRelic collected errors:

```
Error: pollPrice Error: [CKBPriceOracle] sendRequest: request failed, {"message":"timeout of 10000ms exceeded","name":"AxiosError","config":{"transitional":{"silentJSONParsing":true,"forcedJSONParsing":true,"clarifyTimeoutError":false},"transformRequest":[null],"transformResponse":[null],"timeout":10000,"xsrfCookieName":"XSRF-TOKEN","xsrfHeaderName":"X-XSRF-TOKEN","maxContentLength":-1,"maxBodyLength":-1,"env":{},"headers":{"Accept":"application/json, text/plain, */*","User-Agent":"axios/0.27.2"},"method":"get","url":"https://api.coingecko.com/api/v3/simple/price?ids=nervos-network&vs_currencies=usd"},"code":"ECONNABORTED","status":null},Error: [CKBPriceOracle] sendRequest: request failed, {"message":"timeout of 10000ms exceeded","name":"AxiosError","config":{"transitional":{"silentJSONParsing":true,"forcedJSONParsing":true,"clarifyTimeoutError":false},"transformRequest":[null],"transformResponse":[null],"timeout":10000,"xsrfCookieName":"XSRF-TOKEN","xsrfHeaderName":"X-XSRF-TOKEN","maxContentLength":-1,"maxBodyLength":-1,"env":{},"headers":{"Accept":"application/json, text/plain, */*","User-Agent":"axios/0.27.2"},"method":"get","url":"https://api.binance.com/api/v3/trades?symbol=CKBUSDT&limit=1"},"code":"ECONNABORTED","status":null},Error: [CKBPriceOracle] sendRequest: request failed, {"message":"getaddrinfo ENOTFOUND xxx.crypto.com","name":"Error","stack":"Error: getaddrinfo ENOTFOUND xxx.crypto.com\n    at GetAddrInfoReqWrap.onlookup [as oncomplete] (node:dns:71:26)\n    at GetAddrInfoReqWrap.callbackTrampoline (node:internal/async_hooks:130:17)","config":{"transitional":{"silentJSONParsing":true,"forcedJSONParsing":true,"clarifyTimeoutError":false},"transformRequest":[null],"transformResponse":[null],"timeout":10000,"xsrfCookieName":"XSRF-TOKEN","xsrfHeaderName":"X-XSRF-TOKEN","maxContentLength":-1,"maxBodyLength":-1,"env":{},"headers":{"Accept":"application/json, text/plain, */*","User-Agent":"axios/0.27.2"},"method":"get","url":"https://xxx.crypto.com/v2/public/get-trades?instrument_name=CKB_USDT"},"code":"ENOTFOUND","status":null}
    at CKBPriceOracle.pollPrice (/Users/keroro520/cryptape/godwoken-web3/packages/api-server/lib/price-oracle.js:180:34)
    at processTicksAndRejections (node:internal/process/task_queues:96:5)
    at async Promise.all (index 0)
    at async CKBPriceOracle.executePoll (/Users/keroro520/cryptape/godwoken-web3/packages/api-server/lib/price-oracle.js:50:40)
    at async /Users/keroro520/cryptape/godwoken-web3/packages/api-server/lib/base/worker.js:89:24
```

Reply:
updated